spec/examples/running_cluster.rb

```diff
@@ -0,0 +1,118 @@
+shared_examples 'a running pupperware cluster' do
+  require 'timeout'
+  require 'json'
+  require 'rspec/core'
+
+  def puppetserver_health_check(container)
+    %x(docker inspect "#{container}" --format '{{.State.Health.Status}}').chomp
+  end
+
+  def get_puppetdb_state
+    status = %x(docker-compose exec -T puppet curl -s 'http://puppetdb:8080/status/v1/services/puppetdb-status').chomp
+    return JSON.parse(status)['state'] unless status.empty?
+  rescue
+    return ''
+  else
+    return ''
+  end
+
+  def start_puppetserver
+    container = %x(docker-compose ps -q puppet).chomp
+    Timeout::timeout(120) do
+      while container.empty?
+        sleep(1)
+        container = %x(docker-compose ps -q puppet).chomp
+      end
+    end
+  rescue Timeout::Error
+    return ''
+  else
+    status = puppetserver_health_check(container)
+    # puppetserver has a healthcheck, we can let that deal with timeouts
+    while status == 'starting'
+      sleep(1)
+      status = puppetserver_health_check(container)
+    end
+
+    # work around SERVER-2354
+    %x(docker-compose exec puppet puppet config set server puppet)
+
+    return status
+  end
+
+  def run_agent(agent_name)
+    %x(docker run --rm -it --net pupperware_default --name #{agent_name} --hostname #{agent_name} puppet/puppet-agent-ubuntu)
+    return $?
+  end
+
+  def check_report(agent_name)
+    domain = %x(docker-compose exec -T puppet facter domain).chomp
+    body = "{ \"query\": \"nodes { certname = \\\"#{agent_name}.#{domain}\\\" } \" }"
+    out = ''
+    Timeout::timeout(120) do
+      while out.empty?
+        out = %x(docker-compose exec -T puppet curl -s -X POST http://puppetdb:8080/pdb/query/v4 -H 'Content-Type:application/json' -d '#{body}')
+        sleep(1) if out.empty?
+      end
+      JSON.parse(out).first['report_timestamp']
+    end
+  rescue Timeout::Error
+    return ''
+  rescue
+      return ''
+  end
+
+  def clean_certificate(agent_name)
+    domain = %x(docker-compose exec -T puppet facter domain).chomp
+    %x(docker-compose exec -T puppet puppetserver ca clean --certname #{agent_name}.#{domain})
+    return $?
+  end
+
+  def start_puppetdb
+    status = get_puppetdb_state
+    # since pdb doesn't have a proper healthcheck yet, this could spin forever
+    # add a timeout so it eventually returns.
+    Timeout::timeout(600) do
```

Comment:
Was torn about how long to set this timeout to be. Trying to balance CI could take a while to start pdb with not wanting to tie up CI for too long if something went wrong. This will be better once https://github.com/puppetlabs/puppetdb/pull/2619 goes in.

Reply:
Yeah, we can probably ticket some sort of liveness test for the future to make it a little more robust - we'll have to see what happens in Azure Pipelines. I'd expect things to be faster there anyhow, but time will tell.